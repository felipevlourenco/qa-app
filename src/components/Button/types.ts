import { ReactNode } from 'react'

export interface ButtonProp {
  children: ReactNode
  onClick?: () => void
  disabled?: boolean
  variant?: 'primary' | 'secondary'
  type?: 'button' | 'submit' | 'reset'
}
