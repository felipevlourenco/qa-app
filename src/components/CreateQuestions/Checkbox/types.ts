export interface CheckboxProp {
  checked: boolean
  onChange: () => void
}
