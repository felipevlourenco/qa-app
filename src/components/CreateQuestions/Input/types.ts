import { FieldError, UseFormRegister } from 'react-hook-form'
import { AddQuestionAction } from 'store/questions/types'

export interface InputProp {
  id: 'question' | 'answer'
  label: string
  defaultValue?: string
  register: UseFormRegister<AddQuestionAction>
  errors: {
    question?: FieldError | undefined
    answer?: FieldError | undefined
  }
}
