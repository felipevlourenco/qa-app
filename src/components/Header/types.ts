import { ReactNode } from 'react'

export interface HeaderProp {
  children: ReactNode
}
