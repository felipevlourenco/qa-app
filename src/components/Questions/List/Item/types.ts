import { Question } from 'store/questions/types'

export interface QuestionItemProp {
  question: Question
}
