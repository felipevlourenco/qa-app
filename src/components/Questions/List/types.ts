import { Question } from 'store/questions/types'

export interface ListProp {
  questions: Question[]
}
