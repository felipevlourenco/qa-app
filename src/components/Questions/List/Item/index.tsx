import { QuestionItemProp } from 'components/Questions/List/Item/types'
import useToggle from 'hooks/useToggle'
import styles from 'components/Questions/List/Item/styles.module.css'
import editIcon from 'assets/edit.svg'
import deleteIcon from 'assets/delete.svg'
import { useAppDispatch } from 'hooks/useRedux'
import { deleteQuestion, editQuestion } from 'store/questions'
import IconButton from 'components/IconButton'
import { useForm } from 'react-hook-form'
import { AddQuestionAction } from 'store/questions/types'
import Input from 'components/CreateQuestions/Input'
import Button from 'components/Button'

const Item = ({ question }: QuestionItemProp) => {
  const dispatch = useAppDispatch()
  const { isToggled: isCollapsed, toggle } = useToggle()
  const { isToggled: isHovered, toggleOn, toggleOff } = useToggle()
  const { isToggled: isEditing, toggleOn: editOn, toggleOff: editOff } = useToggle()
  const {
    register,
    handleSubmit,
    formState: { errors }
  } = useForm<AddQuestionAction>()

  const deleteHandler = () => {
    dispatch(deleteQuestion({ id: question.id }))
  }

  const onSubmit = handleSubmit((data) => {
    dispatch(editQuestion({ ...data, id: question.id }))
    editOff()
  })

  return (
    <li className={styles.item} onClick={toggle} onMouseEnter={toggleOn} onMouseLeave={toggleOff}>
      <div className={styles.button}>
        {isEditing ? (
          <form className={styles.form} onSubmit={onSubmit}>
            <Input
              id="question"
              label="Question"
              defaultValue={question.question}
              register={register}
              errors={errors}
            />
            <Input id="answer" label="answer" defaultValue={question.answer} register={register} errors={errors} />
            <div className={styles.submit}>
              <Button type="submit">Edit question</Button>
              <Button type="button" variant="secondary" onClick={editOff}>
                Cancel
              </Button>
            </div>
          </form>
        ) : (
          <>
            <span className={styles.question}>{question.question}</span>
            {isCollapsed && <span className={styles.answer}>{question.answer}</span>}
          </>
        )}
      </div>

      {isHovered && (
        <>
          <IconButton icon={editIcon} alt="edit" onClick={editOn} />
          <IconButton icon={deleteIcon} alt="delete" onClick={deleteHandler} />
        </>
      )}
    </li>
  )
}

export default Item
