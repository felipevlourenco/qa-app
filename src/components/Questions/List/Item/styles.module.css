.button {
  margin: 0.5rem;
  background-color: transparent;
  padding: 0.5rem;
  border: 0;
  display: flex;
  flex-direction: column;
  gap: 0.4rem;
}

.question {
  font-size: 1rem;
  font-weight: 600;
}

.question:hover {
  cursor: pointer;
  text-decoration: underline;
}

.answer {
  font-size: 0.9rem;
  font-weight: 200;
}