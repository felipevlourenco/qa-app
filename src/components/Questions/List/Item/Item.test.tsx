import { fireEvent, render } from '@testing-library/react'
import Item from 'components/Questions/List/Item'
import { Question } from 'store/questions/types'
import MockStore from 'test/utils/MockStore'

describe('component/Item', () => {
  const question: Question = {
    id: 1,
    question: 'This is a question',
    answer: 'This is a answer'
  }

  it('should render', () => {
    const { container } = render(
      <MockStore>
        <Item question={question} />
      </MockStore>
    )
    expect(container).toMatchSnapshot()
  })

  it('should answer be visible when question is click', () => {
    const { getByRole, getByText } = render(
      <MockStore>
        <Item question={question} />
      </MockStore>
    )

    fireEvent.click(getByRole('listitem'))
    expect(getByText(/this is a answer/i)).toBeInTheDocument()
  })
})
