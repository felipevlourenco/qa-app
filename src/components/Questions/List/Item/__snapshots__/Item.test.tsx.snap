// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/Item should render 1`] = `
<div>
  <li
    class="item"
  >
    <div
      class="button"
    >
      <span
        class="question"
      >
        This is a question
      </span>
    </div>
  </li>
</div>
`;
